### PR TITLE
Scan JAX-RS sub-resource locators

### LINKED ISSUE #3919: JAX-RS sub resource is not recognised in Quarkus
Register sub resource in the parent resource.

```java
public class PostResource{
    //other methods are omitted here.
    
   @Path("{id}/comments")
    public CommentResource commentResource() {
        return resourceContext.getResource(CommentResource.class);
    }    
} 
```
`CommentResource` is like this, it does not includ a `@Path`.

```java
@RequestScoped
public class CommentResource {}
```

The details and workaround can be found on stackoverflow.
https://stackoverflow.com/questions/57820428/jax-rs-subresource-issue-in-quarkus

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #3919

### CODE REVIEW

File: extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java [hunk 1]
```diff
@@ -262,6 +277,85 @@ public void build(
...
+        Set<DotName> subresourceLocatorTypes = new HashSet<>();
+        for (ClassInfo resourceClass : scannedResources.values()) {
+            ClassInfo clazz = resourceClass;
+            while (clazz != null) {
+                for (MethodInfo method : clazz.methods()) {
+                    if (method.hasAnnotation(ResteasyDotNames.PATH)) {
+                        subresourceLocatorTypes.add(method.returnType().name());
+                    }
+                }
+                if (clazz.superName().equals(DotNames.OBJECT)) {
```
Comment (geoand) [inline extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java]: Does this `if / else` perhaps need to be inverted?

Comment (mkouba) [reply extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java]: Good catch! Thanks.

File: extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/subresource/SuperPingResource.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package io.quarkus.resteasy.test.subresource;
+
+import javax.enterprise.inject.Typed;
+import javax.ws.rs.GET;
+
+// WORKAROUND - resources are beans, without this restriction an ambiguous resolution would occur
```
Comment (mkouba) [inline extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/subresource/SuperPingResource.java]: This is not very user-friendly but it's a result of the current integration design where resources are beans. We had a similar issue with filters: https://github.com/quarkusio/quarkus/issues/3369#issuecomment-520319715. I think we could try to support `InjectionTarget` in ArC but it wouldn't be trivial. So I'm wondering if we should restrict the bean types of all JAX-RS resources, ie. add `@Typed(SuperPingResource.class)` automatically to all resources. @stuartwdouglas @geoand @gsmet @patriot1burke WDYT?

Comment (geoand) [reply extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/subresource/SuperPingResource.java]: Sounds fine to me

**Review by gsmet — DISMISSED** [review]
I added one comment and a question.

**Comment by mkouba** [comment]
This PR should be ready now.

**Comment by gsmet** [comment]
@mkouba unfortunately, we have a conflict, now. Could you rebase? Thanks!

File: extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java [hunk 3]
```diff
@@ -271,6 +287,61 @@ public void build(
...
+                    return;
+                }
+                if (BuiltinScope.isIn(clazz.classAnnotations())) {
+                    // Skip classes annotated with built-in scope
+                    return;
+                }
+                if (clazz.classAnnotation(ResteasyDotNames.PROVIDER) != null) {
+                    Transformation transformation = null;
+                    if (clazz.annotations().containsKey(DotNames.INJECT)
+                            || hasAutoInjectAnnotation(autoInjectAnnotationNames, clazz)) {
```
Comment (mkouba) [inline extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java]: @gwenneg This should fix #4020 for any "auto-inject" annotation.

Comment (gwenneg) [reply extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java]: @mkouba Thanks!

**Comment by mkouba** [comment]
@gsmet rebased.

**Review by manovotn — APPROVED** [review]

**Pull request merged**